### PR TITLE
8306882: (fs) Path.toRealPath(LinkOption.NOFOLLOW_LINKS) fails when "../../" follows a link

### PR DESCRIPTION
**Pull request by bplb**
Modify `Path.toRealPath` such that it does not collapse links such as "link/.." or "link/../.." when `LinkOption.NOFOLLOW_LINKS` is specified or, in some cases, not.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306882](https://bugs.openjdk.org/browse/JDK-8306882): (fs) Path.toRealPath(LinkOption.NOFOLLOW_LINKS) fails when "../../" follows a link (**Bug** - P4)

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [75db4996](https://git.openjdk.org/jdk/pull/15397/files/75db49960921293dd6b50c96e0961475c8ec2af4)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/15397/head:pull/15397` \
`$ git checkout pull/15397`

Update a local copy of the PR: \
`$ git checkout pull/15397` \
`$ git pull https://git.openjdk.org/jdk.git pull/15397/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 15397`

View PR using the GUI difftool: \
`$ git pr show -t 15397`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/15397.diff">https://git.openjdk.org/jdk/pull/15397.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/15397#issuecomment-1689129534)

### CODE REVIEW

**Comment by bplb** [comment]
On Windows, the function `GetFullPathName()` does not access the file system and so is susceptible to collapsing links resulting in a path to a non-existent file. The path is examined to see whether it is one for which `GetFullPathName` would generate an incorrect result. If that is **not** the case, then `GetFullPathName` is used as before and there is no effective change.  So this modification should affect only cases which would otherwise generate an incorrect result or fail.

File: src/java.base/windows/classes/sun/nio/fs/WindowsLinkSupport.java [hunk 1]
```diff
@@ -190,20 +190,92 @@ static String getRealPath(WindowsPath input, boolean resolveLinks)
...
-            } catch (WindowsException x) {
-                x.rethrowAsIOException(input);
+            // The Windows function GetFullPathName() does not access the
+            // file system, hence is susceptible to removing symbolic links
+            // which are followed by the special name ".." whether or not
+            // links are being followed. Examine the path element-by-element
+            // to determine whether this would occur, and if so use the
+            // full path manually derived in the process.
+            String fullpath = null;
+            if (path.contains("..")) {
```
Comment (turbanoff) [inline src/java.base/windows/classes/sun/nio/fs/WindowsLinkSupport.java:211]: I suggest to extract code for this case to a separate method. `sun.nio.fs.WindowsLinkSupport#getRealPath` is already too big.

**Comment by AlanBateman** [comment]
The changes to the Windows implementation of toRealPath(LinkOption.NOFOLLOW_LINKS) are going to take time to review. 

Once suggestion that would help set the groundwork is to move the toRealPath tests to their own JUnit test. It's just historically that they were initially put in Misc.java. Also I think it would be useful to do the change to the Unix implementation separately, I think that needs a summary in the first place to understand why that change is needed.

File: src/java.base/windows/classes/sun/nio/fs/WindowsLinkSupport.java (hunk 1 above)
Comment (bplb) [reply src/java.base/windows/classes/sun/nio/fs/WindowsLinkSupport.java:211]: Sounds like a good idea.

**Comment by bplb** [comment]
> Also I think it would be useful to do the change to the Unix implementation separately, I think that needs a summary in the first place to understand why that change is needed.

The issue was filed against the Unix implementation, so if the Unix implementation is separated from Windows then I think the Windows content should be moved to a new PR.

**Comment by bplb** [comment]
> Once suggestion that would help set the groundwork is to move the toRealPath tests to their own JUnit test.

See [JDK-8315241](https://bugs.openjdk.org/browse/JDK-8315241).

**Comment by bplb** [comment]
> I think that [the Unix implementation] needs a summary in the first place to understand why that change is needed.

I can add a comment in the source, but here's some illustrative behavior using `jshell`. It uses some files as:
```
./dir/subdir
./aaa@ -> dir/subdir
./out.txt
```
and attempts to convert `aaa/../../out.txt` to a real path.

**master**
```
jshell> Path p = Path.of("aaa/../../out.txt")
p ==> aaa/../../out.txt

jshell> p.toRealPath()
$2 ==> /Users/bpb/dev/bugs/jdk/Path-toRealPath-8306882/out.txt                                        

jshell> p.toRealPath(LinkOption.NOFOLLOW_LINKS)
|  Exception java.nio.file.NoSuchFileException: /Users/bpb/dev/bugs/jdk/Path-toRealPath-8306882/aaa/out.txt
|        at UnixException.translateToIOException (UnixException.java:92)
|        at UnixException.rethrowAsIOException (UnixException.java:106)
|        at UnixException.rethrowAsIOException (UnixException.java:111)
|        at UnixPath.toRealPath (UnixPath.java:927)
|        at (#3:1)
```
**patch**
```
jshell> Path p = Path.of("aaa/../../out.txt")
p ==> aaa/../../out.txt

jshell> p.toRealPath()
$2 ==> /Users/bpb/dev/bugs/jdk/Path-toRealPath-8306882/out.txt

jshell> p.toRealPath(LinkOption.NOFOLLOW_LINKS)
$3 ==> /Users/bpb/dev/bugs/jdk/Path-toRealPath-8306882/aaa/../../out.txt
```
As can be seen, when attempting to resolve the path without following links, the result is a non-existent file.

**Comment by bplb** [comment]
The test and Windows changes reverted in 75db49960921293dd6b50c96e0961475c8ec2af4 will be addressed in #15474 and a separate issue to be filed, respectively.

**Comment by bplb** [comment]
> [...] I assume the work on converting the test to a standalone JUnit test will expand the test coverage to include this scenario.

It will, and the test will initially have to be excluded on Windows until [JDK-8315273](https://bugs.openjdk.org/browse/JDK-8315273) is fixed.

**Pull request closed without merging**